src/shogun/kernel/Kernel.cpp

```diff
@@ -109,15 +109,15 @@ bool CKernel::init(CFeatures* l, CFeatures* r)
 	}
 	ASSERT(l->get_feature_type()==r->get_feature_type())
 
-	//remove references to previous features
-	remove_lhs_and_rhs();
-
 	SG_REF(l);
 	if (l==r)
 		lhs_equals_rhs=true;
 	else // l!=r
 		SG_REF(r);
 
+	//remove references to previous features
+	remove_lhs_and_rhs();
+
```

Comment:
Factories create objects with refcount = 0. When init is called with two features, and then is called again with a feature in common, calling "remove_lhs_and_rhs" first  results in the feature object being deleted, causing segfaults.

Reply:
why don't factories reference the objects before returning them?

Reply:
Was recently rly changed, Viktor can comment
I would just add a workaround as soon we will use shared ptr

Reply:
That’s because swig increases the refcount instead. So we ended with refcount of 2, and that leaked memory

Reply:
@gf712 I checked it on my machine. The kernel I created had a refcount of zero.

Reply:
the feature object i mean, sorry

Reply:
@gf712 I checked other objects. You're right, swig does increase the reference count for all except string_features

Reply:
@theartful is there something missing in https://github.com/shogun-toolbox/shogun/blob/6fb8b6f613a6aada95cd9f53b588111ccff6279b/src/interfaces/swig/factory.i#L19

Reply:
signature is different, i will check